README.org
* Introduction

[[https://travis-ci.org/schwehr/libais][https://travis-ci.org/schwehr/libais.svg?branch=master]]
[[https://scan.coverity.com/projects/5519][https://scan.coverity.com/projects/5519/badge.svg]]
[[https://codeclimate.com/github/schwehr/libais][https://codeclimate.com/github/schwehr/libais/badges/gpa.svg]]
[[http://badge.fury.io/py/libais][https://badge.fury.io/py/libais.svg]]

Library for decoding maritime Automatic Identification System messages.

* See Also

http://en.wikipedia.org/wiki/Automatic_Identification_System

Other open source AIS projects:

- http://en.wikipedia.org/wiki/Gpsd
- http://www.freeais.org/ - does this actually decode AIS?
- http://github.com/DaMSA/AisLib
- http://github.com/schwehr/noaadata
- https://github.com/schwehr/ais-areanotice-py
- https://github.com/OpenCPN/OpenCPN
- https://github.com/bcl/aisparser

* Building

** Building with legacy Makefile

#+BEGIN_SRC sh
make -f Makefile-custom test
#+END_SRC

** Building with python

#+BEGIN_SRC sh
python setup.py build
python setup.py install
#+END_SRC

** Building with CMake

#+BEGIN_SRC sh
cmake .
make
#+END_SRC

* Usage

#+BEGIN_SRC sh
ipython
#+END_SRC

There are two interfaces to libais, one high-level iterator based one
and a low-level fast C++ only one. The iterator based interface is
accessed the following way:

#+BEGIN_SRC python
import ais.stream
with open("file.nmea") as f:
    for msg in ais.stream.decode(f):
        print msg
#+END_SRC

To use the low-level C++ interface directly, you need to handle multi-line messages and padding yourself:

#+BEGIN_SRC python
import ais
ais.decode('15PIIv7P00D5i9HNn2Q3G?wB0t0I', 0)
ais.decode('402u=TiuaA000r5UJ`H4`?7000S:', 0)
ais.decode('55NBjP01mtGIL@CW;SM<D60P5Ld000000000000P0`<3557l0<50@kk@K5h@00000000000', 0)
#+END_SRC

There is also support for converting parsed messages to the structure
output by GPSD / gpsdecode. For full compatibility, you have to write
the resulting message dictionaries to a file with json.dump() and add
a newline after each message.

#+BEGIN_SRC python
import ais.stream
import json
import ais.compatibility.gpsd

with open("infile.nmea") as inf:
    with open("outfile.gpsd") as outf:
        for msg in ais.stream.decode(f):
            gpsdmsg = ais.compatibility.gpsd.mangle(msg)
            json.dump(gpsdmsg, outf)
            outf.write("\n")
#+END_SRC

* AIS Specification Documents

- ITU-1371, ITU-1371-{1,2,3,4]
- http://www.e-navigation.nl/asm
- IMO Circ 236
- IMO Circ 289
- EU RIS

* Hacking

The C++ code was switched to the Google style in November, 2012.
Indenting should be by 2 spaces.

http://google-styleguide.googlecode.com/svn/trunk/cpplint/


Changelog.md
Changelog and release history for libais
========================================

0.17 - 2018-01-17
-----------------

* Fixes to 10, 12, 16, 26
* Fixes to 6:1:0, 6:1:3, 6:1:4, 6:1:5, 8:1:0, 8:367:22, 8:200:24
* Add 8:200:21, 8:200:22
* Cleanup nmea_messages.py
* Use a logger in python
* Factor out ais_bitset.cpp
* Stop calling ParseNmeaPayload multiple times per message
* Stay far away from -ffast-math
* Fixes for CMake builds

0.16 - 2015-11-12
-----------------

* Make setup.py more functional for development and testing
* Version number is now in ais/__init__.py
* Travis-CI container testing
* Converted README.org to README.rst
* Converted Changelog.html to Changelog.md
* Add ais/io.py
* Add ais/nmea_messages.py for non-AIS messages
* Some improvements to ais/stream and aisdecode, but they are now deprecated
* TAG Block and USCG parsing converts message parts to python types
* Moved python code out of the top level directory
* Remove legacy SQL, Nagios and NAIS tools
* Fix initialization of ais12, ais16, ais17, ais18, ais8_1_22, ais8_366_22
* Fix memory leak in python handling of ais20
* Add C++ decode_body and vdm handler

0.15 - 2015-06-16
-----------------

* Added libais namespace
* Started using clang-format
* Fix Ais18 carrier sense (CS) messages
* Use AisBitset for bit decoding
* Fix AIS_STATUS_STRINGS
* Use std::array for fixed sized arrays
* Update DAC enum with new values
* Added ais/compatibility/gpsd.py
* Added ais/stream (deprecated - please do not use)
* Closed github issues 27, 31, 34, 36, 38, 42, 45, 46, 47, 51, 52, 53, 56, and 60
* Added ais/nmea.py, ais/tag_block.py, and ais/util_test.py
* Rearranged the tree structure to have src/... for c++ and {ais,test}/ for python
* Use more c++11: NULL -&gt; nullptr
* Add 8:367:22 Area Notice
* python setup.py test mostly works
* At least one test for all top levels except msg 6.
* Add Travis CI testing

0.14 - 2014-04-22
-----------------

* Switch license from LGPL 3+ to Apache 2.0
* Msg 8:366:22 bit count check fixed
* Fixed spare bit calculation in msg 8
* Added DAC country codes and FI message ids enums
* Fixed error tracking in msg 6 and 8 sub-messages

0.13 - 2012-11-18
-----------------

* Switch to the [Google C++ style guide](http://code.google.com/p/google-styleguide/) and [`cpplint.py`](http://google-styleguide.googlecode.com/svn/trunk/cpplint/)
* Lots of small bugs found in the code review process
* Message constructors now start as status = AIS_UNINITIALIZED and set to AIS_OK when decoding is done
* Switched to using initializers to call parent initializers in C++
* Removed a lot of duplicate and unused code. AIS_ERR_WRONG_MSG_TYPE removed.  More still needs to be removed.
* Rewrote the C++ NMEA parsing functions: GetBody, GetNthField, GetPad and Split.
* Switch to using asserts for checks that imply coding errors within the library.


0.12 - 2012-11-05
-----------------

* Fix bit count bugs in 8_1_14, 8_1_15, 8_1_27
* Rewrote nth_field.  Added split and delimiters
* Folded in 366 header to ais.h
* Lots and lots of style cleanup
* Use std::foo, remove std:: from code
* Documentation for Msg 17
* Testing of Msg 20


0.11 - 2012-10-29
-----------------

* New release because a large binary went out in 0.10


0.10 - 2012-10-29
-----------------

* Add a test directory and test of all top level msgs except 20 in python
* Begin cleanup of test_libais.cpp
* Almost all decoders require pad bits now
* linted - lots of formatting changes
* Start implementing Msg 17 GNSS differential corrections
* Convert FIX to TODO and put (schwehr) after each to assign them to myself.
* remove bool casting of bitset[offset]
* Implemented 8 1 26
* Clean up c++ logical operators.  Do not use and, or, and not
* Message 24 needed pad.  Fixed
* Removed print()
* remove almost all cout/cerr that were not in print()
* Remove lots of dead code
* Pass pad into ais.decode in python, but handle without


0.9 - 2012-10-19
-----------------

* ais.decode now requires the pad bits in python
* Added RIS 8_200_{10,23,24,40,55}
* Implemented the rest of Circ 236 BBM
* Implemented all Circ 289 messages except ABM route and BBM env sensors
* Implemented AIS messages 6, 9, 10, 12, 15-17, 20-22, 23, 25-27. Still payload work to do.
* Imported rolker's CMakeList.txt


0.8 - 2012-05-12
-----------------

* MANIFEST.in now has VERSION
* Implemented AIS messages 7, 13, 14, 18, 19, 24


0.7 - 2012-04-30
-----------------

* Added MANIFEST.in
* setup.py compliant with pypi
* Added AIS area/zone messages 8:1:22 and 8:366:22


0.6 - 2010-06-17
-----------------

* ais21.cpp: new file - AtoN status
* ais.h: fix CHECKPOINT for emacs 23
* ais.h: Proper inheritance from AisMsg of message_id, repead_id and mmsi
* ais.h: started trying to define 8_366_34 - zone msg
* nagios_pg_ais.py: new file - monitor db statios with nagios ssh or snmp


0.5 - 2010-05-25
-----------------

* Still a lot of untested/unimplemented messages
* docs: Include ESR's AIVDM.txt with permission
* docs: MID / DAC / MMSI prefixes now listed in mid.csv
* docs: dac/fi list
* docs: More notes for message designers
* Added msg 8 - 1:11 - IMO Met/Hydro
* Added AIS msg 9 - SAR Position
* nais2py.py: Added try except wrapper on ProcessingThread.  Also try to track one off error found on call with timestamp converting to float
* send_data.py: new file for testing
* nais2py.py: LineQueue now has a custom drop handler for too many lines waiting.  Could be better.  
* nais2py.py: Added threaded network interface.  Seeing the network side overwhelm the processing thread
* nais2py.py: Added response_class handling to VesselNames.  Can be preloaded.  Allows periodic name dump
* nais2py.py: Added ENABLE_DB flag to try runs without database execute commands.  Faster debugging
* vessels.csv: new file - example preloading of vessel names as response ships


0.4 - 2010-05-11
-----------------

* nais2py.py: Started providing a command line interface
* nais2py.py: Added PositionCache class

ais6.cpp
// Address Binary Message (ABM) 6

#include <iomanip>
#include <cmath>

#include "ais.h"

Ais6::Ais6(const char *nmea_payload, const size_t pad)
    : AisMsg(nmea_payload, pad), seq(0), mmsi_dest(0), retransmit(false),
      spare(0), dac(0), fi(0) {
  assert(message_id == 6);

  // TODO(olafsinram): 46 or rather 56??
  const int payload_len = num_bits - 46;  // in bits w/o DAC/FI
  if (num_bits < 88 || payload_len < 0 || payload_len > 952) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(38);
  seq = bs.ToUnsignedInt(38, 2);
  mmsi_dest = bs.ToUnsignedInt(40, 30);
  retransmit = !bs[70];
  spare = bs[71];
  dac = bs.ToUnsignedInt(72, 10);
  fi = bs.ToUnsignedInt(82, 6);
}

Ais6_1_0::Ais6_1_0(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), ack_required(false), msg_seq(0),
      spare2(0) {

  assert(dac == 1);
  assert(fi == 0);

  if (num_bits < 88 || num_bits > 936) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;  // TODO(schwehr): what is the real max size?
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  ack_required = bs[88];
  msg_seq = bs.ToUnsignedInt(89, 11);

  const size_t text_size = 6 * ((num_bits - 100) / 6);
  const size_t spare2_size = num_bits - 100 - text_size;
  text =  bs.ToString(100, text_size);

  if (!spare2_size)
    spare2 = 0;
  else
    spare2 = bs.ToUnsignedInt(100 + text_size, spare2_size);

  status = AIS_OK;
}

Ais6_1_1::Ais6_1_1(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), ack_dac(0), msg_seq(0), spare2(0) {
  assert(dac == 1);
  assert(fi == 1);

  if (num_bits != 112) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  ack_dac = bs.ToUnsignedInt(88, 10);
  msg_seq = bs.ToUnsignedInt(98, 11);
  spare2 = bs.ToUnsignedInt(109, 3);

  status = AIS_OK;
}

Ais6_1_2::Ais6_1_2(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), req_dac(0), req_fi(0) {
  assert(dac == 1);
  assert(fi == 2);

  if (num_bits != 104) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  req_dac = bs.ToUnsignedInt(88, 10);
  req_fi = bs.ToUnsignedInt(98, 6);

  status = AIS_OK;
}

// IFM 3: Capability interrogation - OLD ITU 1371-1
Ais6_1_3::Ais6_1_3(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), req_dac(0), spare2(0) {
  assert(dac == 1);
  assert(fi == 3);

  if (num_bits != 104) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  req_dac = bs.ToUnsignedInt(88, 10);
  spare2 = bs.ToUnsignedInt(94, 6);

  status = AIS_OK;
}

// IFM 4: Capability reply - OLD ITU 1371-4
// TODO(schwehr): WTF?  10 + 128 + 6 == 80  Is this 168 or 232 bits?
Ais6_1_4::Ais6_1_4(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), ack_dac(0), capabilities(),
      cap_reserved(), spare2(0) {
  assert(dac == 1);
  assert(fi == 4);

  // TODO(schwehr): num_bits for 6_1_4.  226 bits?
  if (num_bits != 232) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  ack_dac = bs.ToUnsignedInt(88, 10);
  for (size_t cap_num = 0; cap_num < 128/2; cap_num++) {
    size_t start = 98 + cap_num * 2;
    capabilities[cap_num] = bs[start];
    cap_reserved[cap_num] = bs[start + 1];
  }
  // spare2 = bs.ToUnsignedInt(226, 6);  // OR NOT
  // TODO(schwehr): add in the offset of the dest mmsi

  status = AIS_OK;
}

// IMO Circ 289 - Dangerous cargo
// See also Circ 236
Ais6_1_12::Ais6_1_12(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), utc_month_dep(0), utc_day_dep(0),
      utc_hour_dep(0), utc_min_dep(0), utc_month_next(0),
      utc_day_next(0), utc_hour_next(0), utc_min_next(0),
      un(0), value(0), value_unit(0), spare2(0) {
  assert(dac == 1);
  assert(fi == 12);

  if (num_bits != 360) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  // TODO(schwehr): add in the offset of the dest mmsi

#if 0
  bs.SeekTo(56);
  last_port = bs.ToString(56, 30);
  utc_month_dep = bs.ToUnsignedInt(86, 4);
  utc_day_dep = bs.ToUnsignedInt(90, 5);
  utc_hour_dep = bs.ToUnsignedInt(95, 5);
  utc_min_dep = bs.ToUnsignedInt(100, 6);
  next_port = bs.ToString(106, 30);
  utc_month_next = bs.ToUnsignedInt(136, 4);  // estimated arrival
  utc_day_next = bs.ToUnsignedInt(140, 5);
  utc_hour_next = bs.ToUnsignedInt(145, 5);
  utc_min_next = bs.ToUnsignedInt(150, 6);
  main_danger = bs.ToString(156, 120);
  imo_cat = bs.ToString(276, 24);
  un = bs.ToUnsignedInt(300, 13);
  value = bs.ToUnsignedInt(313, 10);  // TODO(schwehr): units
  value_unit = bs.ToUnsignedInt(323, 2);
  spare = bs.ToUnsignedInt(325, 3);
  // 360
#endif

  status = AIS_OK;
}

// 6_1_13 Does not exist

// IMO Circ 289 - Tidal Window
// See also Circ 236
Ais6_1_14::Ais6_1_14(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), utc_month(0), utc_day(0) {
  // TODO(schwehr): untested - no sample of the correct length yet
  assert(dac == 1);
  assert(fi == 14);

  if (num_bits != 376) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  utc_month = bs.ToUnsignedInt(88, 4);
  utc_day = bs.ToUnsignedInt(92, 5);

  for (size_t window_num = 0; window_num < 3; window_num++) {
    Ais6_1_14_Window w;
    const size_t start = 97 + window_num * 93;
    w.y = bs.ToInt(start, 27) / 600000.;
    w.x = bs.ToInt(start + 27, 28) / 600000.;

    w.utc_hour_from = bs.ToUnsignedInt(start + 55, 5);
    w.utc_min_from = bs.ToUnsignedInt(start + 60, 6);
    w.utc_hour_to = bs.ToUnsignedInt(start + 66, 5);
    w.utc_min_to = bs.ToUnsignedInt(start + 71, 6);
    w.cur_dir = bs.ToUnsignedInt(start + 77, 9);
    w.cur_speed  = bs.ToUnsignedInt(start + 86, 7) / 10.;

    windows.push_back(w);
  }

  status = AIS_OK;
}

// IMO Circ 289 - Clearance time to enter port
Ais6_1_18::Ais6_1_18(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), link_id(0), utc_month(0), utc_day(0),
      utc_hour(0), utc_min(0), spare2() {
  assert(dac == 1);
  assert(fi == 18);

  if (num_bits != 360) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  link_id = bs.ToUnsignedInt(88, 10);
  utc_month = bs.ToUnsignedInt(98, 4);
  utc_day = bs.ToUnsignedInt(102, 5);
  utc_hour = bs.ToUnsignedInt(107, 5);
  utc_min = bs.ToUnsignedInt(112, 6);
  port_berth = bs.ToString(118, 120);
  dest = bs.ToString(238, 30);
  x = bs.ToInt(268, 25) / 60000.;
  y = bs.ToInt(293, 24) / 60000.;
  spare2[0] = bs.ToUnsignedInt(317, 32);
  spare2[1] = bs.ToUnsignedInt(349, 11);

  status = AIS_OK;
}

// IMO Circ 289 - Berthing data
Ais6_1_20::Ais6_1_20(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), link_id(0), length(0), depth(0.0),
      position(0), utc_month(0), utc_day(0), utc_hour(0), utc_min(0),
      services_known(false), services() {
  assert(dac == 1);
  assert(fi == 20);

  if (num_bits != 360) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  link_id = bs.ToUnsignedInt(88, 10);
  length = bs.ToUnsignedInt(98, 9);
  depth = bs.ToUnsignedInt(107, 8);
  position = bs.ToUnsignedInt(115, 3);
  utc_month = bs.ToUnsignedInt(118, 4);
  utc_day = bs.ToUnsignedInt(122, 5);
  utc_hour = bs.ToUnsignedInt(127, 5);
  utc_min = bs.ToUnsignedInt(132, 6);
  services_known = bs[138];
  for (size_t serv_num = 0; serv_num < 26; serv_num++) {
    // TODO(schwehr): const int val = bs.ToUnsignedInt(139 + 2*serv_num, 2);
    services[serv_num]
        = static_cast<int>(bs.ToUnsignedInt(139 + 2*serv_num, 2));
  }
  name = bs.ToString(191, 120);
  x = bs.ToInt(311, 25);
  y = bs.ToInt(336, 24);

  status = AIS_OK;
}

Ais6_1_25_Cargo::Ais6_1_25_Cargo()
    : code_type(0), imdg_valid(false), imdg(0), spare_valid(false), spare(0),
      un_valid(false), un(0), bc_valid(false), bc(0), marpol_oil_valid(false),
      marpol_oil(0), marpol_cat_valid(false), marpol_cat(0) {}

// IMO Circ 289 - Dangerous cargo indication 2
// See also Circ 236
Ais6_1_25::Ais6_1_25(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), amount_unit(0), amount(0) {
  assert(dac == 1);
  assert(fi == 25);

  // TODO(schwehr): verify multiple of the size of cargos + header
  //   or padded to a slot boundary
  // Allowing a message with no payloads
  // TODO(schwehr): (num_bits - 100) % 17 != 0) is okay
  if (num_bits < 100 || num_bits > 576) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }
  if ((num_bits - 100) % 17 != 0) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  amount_unit = bs.ToUnsignedInt(88, 2);
  amount = bs.ToUnsignedInt(90, 10);
  const size_t total_cargos = static_cast<int>(floor((num_bits - 100) / 17.));
  for (size_t cargo_num = 0; cargo_num < total_cargos; cargo_num++) {
    Ais6_1_25_Cargo cargo;
    const size_t start = 100 + 17*cargo_num;
    cargo.code_type = bs.ToUnsignedInt(start, 4);

    // TODO(schwehr): Is this the correct behavior?
    switch (cargo.code_type) {
      // No 0
      case 1:  // IMDG Code in packed form
        cargo.imdg = bs.ToUnsignedInt(start + 4, 7);
        cargo.imdg_valid = true;
        cargo.spare = bs.ToUnsignedInt(start + 11, 6);
        cargo.spare_valid = true;
        break;
      case 2:  // IGC Code
        cargo.un = bs.ToUnsignedInt(start + 4, 13);
        cargo.un_valid = true;
        break;
      case 3:  // BC Code
        cargo.bc = bs.ToUnsignedInt(start + 4, 3);
        cargo.bc_valid = true;
        cargo.imdg = bs.ToUnsignedInt(start + 7, 7);
        cargo.imdg_valid = true;
        cargo.spare = bs.ToUnsignedInt(start + 14, 3);
        cargo.spare_valid = true;
        break;
      case 4:  // MARPOL Annex I
        cargo.marpol_oil = bs.ToUnsignedInt(start + 4, 4);
        cargo.marpol_oil_valid = true;
        cargo.spare = bs.ToUnsignedInt(start + 8, 9);
        cargo.spare_valid = true;
        break;
      case 5:  // MARPOL Annex II IBC
        cargo.marpol_cat = bs.ToUnsignedInt(start + 4, 3);
        cargo.marpol_cat_valid = true;
        cargo.spare = bs.ToUnsignedInt(start + 7, 10);
        cargo.spare_valid = true;
        break;
      // 6: Regional use
      // 7: 7-15 reserved for future
      default:
        break;  // Just push in an all blank record?
    }
    cargos.push_back(cargo);
  }

  status = AIS_OK;
}

// TODO(schwehr): 6_1_28 - Modify 8_1_28 once that is debugged

Ais6_1_32_Window::Ais6_1_32_Window()
    : from_utc_hour(0), from_utc_min(0), to_utc_hour(0), to_utc_min(0),
      cur_dir(0), cur_speed(0.0) {}

// IMO Circ 289 - Tidal window
// See also Circ 236
Ais6_1_32::Ais6_1_32(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), utc_month(0), utc_day(0) {
  assert(dac == 1);
  assert(fi == 32);

  // TODO(schwehr): might get messages with not all windows
  if (num_bits != 350) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  utc_month = bs.ToUnsignedInt(88, 4);
  utc_day = bs.ToUnsignedInt(92, 5);

  for (size_t window_num = 0; window_num < 3; window_num++) {
    Ais6_1_32_Window w;
    const size_t start = 97 + 88*window_num;
    w.x = bs.ToInt(start, 25) / 60000.;
    w.y = bs.ToInt(start + 25, 24) / 60000.;
    w.from_utc_hour = bs.ToUnsignedInt(start + 49, 5);
    w.from_utc_min = bs.ToUnsignedInt(start + 54, 6);
    w.to_utc_hour = bs.ToUnsignedInt(start + 60, 5);
    w.to_utc_min = bs.ToUnsignedInt(start + 65, 6);
    w.cur_dir = bs.ToUnsignedInt(start + 71, 9);
    w.cur_speed = bs.ToUnsignedInt(start + 80, 8) / 10.;
    windows.push_back(w);
  }

  status = AIS_OK;
}

// IFM 40: people on board - OLD ITU 1371-4
Ais6_1_40::Ais6_1_40(const char *nmea_payload, const size_t pad)
    : Ais6(nmea_payload, pad), persons(0), spare2(0) {
  assert(dac == 1);
  assert(fi == 40);

  if (num_bits != 104) {
    status = AIS_ERR_BAD_BIT_COUNT;
    return;
  }

  AisBitset bs;
  const AIS_STATUS r = bs.ParseNmeaPayload(nmea_payload, pad);
  if (r != AIS_OK) {
    status = r;
    return;
  }

  bs.SeekTo(88);
  persons = bs.ToUnsignedInt(88, 13);
  spare2 = bs.ToUnsignedInt(101, 3);

  status = AIS_OK;
}
